src/Compilers/CSharp/Portable/Binder/Semantics/OverloadResolution/MethodTypeInference.cs

```diff
@@ -3153,7 +3153,14 @@ private static bool IsReallyAType(TypeSymbol? type)
 
         private static void GetAllCandidates(Dictionary<TypeWithAnnotations, TypeWithAnnotations> candidates, ArrayBuilder<TypeWithAnnotations> builder)
         {
-            builder.AddRange(candidates.Values);
+            // Not using builder.AddRange here because the dictionary values enumerator is a struct, and calling AddRange will have to box the enumerator.
+            // Instead, we increase the builder capacity, then loop over the values.
+            // Also, we don't access candidates.Values to avoid realizing the Dictionary's internal ValueCollection (it's created lazily when Values is accessed)
+            builder.Count += candidates.Count;
+            foreach (var (_, value) in candidates)
```

Comment:
>candidates [](http://example.com/codeflow?start=39&length=10)

Is there a specific reason to enumerate pair rather than ```.Values ``` collection?

Reply:
.Values allocates.  While the kvp enumerator is a struct enumerator.  :'(

Reply:
Indeed. That is also mentioned in the code comment to avoid someone refactoring this in future to use `.Values`:

> // Also, we don't access candidates.Values to avoid realizing the Dictionary's internal ValueCollection (it's created lazily when Values is accessed)